include/thousandeyes/futures/PollingExecutor.h

```diff
@@ -106,61 +109,58 @@ class PollingExecutor :
                     waitables_.pop();
                 }
 
-                bool ready = true;
-                std::exception_ptr error = nullptr;
-
                 try {
-                    ready = w->wait(q_);
+                    if (!w->wait(q_)) {
+                        std::lock_guard<std::mutex> lock(mutex_);
+                        waitables_.push(std::move(w));
+                        continue;
+                    }
+
+                    dispatch_(std::move(w), nullptr);
                 }
                 catch (...) {
-                    error = std::current_exception();
+                    dispatch_(std::move(w), std::current_exception());
                 }
-
-                if (!ready) {
-                    std::lock_guard<std::mutex> lock(mutex_);
-
-                    waitables_.push(std::move(w));
-                    continue;
-                }
-
-                // Using shared_ptr to enable copy-ability of the lambda, otherwise the
-                // dispatchFunc_ would not be able to accept it as function<void()>
-                std::shared_ptr<Waitable> wShared = std::move(w);
-                (*dispatchFunc_)([w=std::move(wShared), error=std::move(error)]() {
-                    w->dispatch(error);
-                });
             }
         });
     }
 
     void stop() override final
     {
-        bool wasActive;
         std::queue<std::unique_ptr<Waitable>> pending;
         {
             std::lock_guard<std::mutex> lock(mutex_);
 
-            wasActive = active_;
             active_ = false;
-
             pending.swap(waitables_);
         }
 
-        if (wasActive) {
-            pollFunc_.reset();
-            dispatchFunc_.reset();
+        while (!pending.empty()) {
+            cancel_(std::move(pending.front()), "Executor stoped");
+            pending.pop();
         }
 
-        if (!pending.empty()) {
-            auto error = std::make_exception_ptr(WaitableWaitException("Executor stoped"));
-            do {
-                pending.front()->dispatch(error);
-                pending.pop();
-            } while (!pending.empty());
-        }
+        pollFunc_.reset();
+        dispatchFunc_.reset();
     }
 
 private:
+    inline void dispatch_(std::unique_ptr<Waitable> w, std::exception_ptr error)
+    {
+        // Using shared_ptr to enable copy-ability of the lambda, otherwise the
+        // dispatchFunc_ would not be able to accept it as function<void()>
+        std::shared_ptr<Waitable> wShared = std::move(w);
+        (*dispatchFunc_)([w=std::move(wShared), error=std::move(error)]() {
```

Comment:
I'm wondering, if you had `Invoker::operator()(function<void()>&& f)`, would that avoid the `shared_ptr`?

Reply:
No because `std::function<>` is copyable by design. It's possible to implement a move-only `std::function` equivalent but that would be a different type altogether. [See also here](https://stackoverflow.com/questions/25330716/move-only-version-of-stdfunction).